### PR TITLE
Use the array pool in Json writer

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues

* Use the array pool in Json writer

### Description

* Use the array pool in Json writer

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Comment by xuzhg** [comment]
@ysanghi Would you please also take a look this PR? thanks.

File: src/Microsoft.OData.Core/Buffers/BufferUtils.cs [hunk 1]
```diff
@@ -23,12 +23,23 @@ internal static class BufferUtils
         /// <returns>The initialized buffer</returns>
         public static char[] InitializeBufferIfRequired(char[] buffer)
         {
-            if (buffer == null)
+            return InitializeBufferIfRequired(null, buffer);
```
Comment (raheph) [inline src/Microsoft.OData.Core/Buffers/BufferUtils.cs]: Why  do we need this method? can't we do something like this?
if (buffer == null) {
       return  RentFromBuffer(null, buffer);
}
return buffer;

Comment (xuzhg) [reply src/Microsoft.OData.Core/Buffers/BufferUtils.cs]: That's avoiding a duplicated code. At later, if we change something, we only change the second method, don't need to change all methods. Thanks.

File: src/Microsoft.OData.Core/Json/JsonWriter.cs [hunk 2]
```diff
@@ -392,6 +398,7 @@ public void WriteRawValue(string rawValue)
         /// </summary>
         public void Flush()
         {
+            BufferUtils.ReturnToBuffer(this.ArrayPool, this.buffer);
             this.writer.Flush();
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/Json/JsonWriter.cs]: Should we return to pool on Flush or Dispose?

Comment (xuzhg) [reply src/Microsoft.OData.Core/Json/JsonWriter.cs]: @mikepizzo I changed to use the "Dispose".

File: src/Microsoft.OData.Core/Buffers/BufferUtils.cs (hunk 1 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/Buffers/BufferUtils.cs]: I'm okay with that pattern.

---
In reply to: [265697058](https://github.com/OData/odata.net/pull/1420#discussion_r265697058) [](ancestors = 265697058)

**Pull request merged**